src/replication/log_generator.cpp

```diff
@@ -37,311 +37,508 @@ namespace cubreplication
+    assert (p_lsa != NULL);
 
-    for (auto repl_obj_it = m_pending_to_be_added.begin (); repl_obj_it != m_pending_to_be_added.end (); ++repl_obj_it)
+    for (auto repl_obj_it = m_pending_to_be_added.begin(); repl_obj_it != m_pending_to_be_added.end(); ++repl_obj_it)
+    {
+      changed_attrs_row_repl_entry *repl_obj = *repl_obj_it;
+      if (repl_obj->compare_inst_oid(inst_oid))
       {
-	changed_attrs_row_repl_entry *repl_obj = *repl_obj_it;
-	if (repl_obj->compare_inst_oid (inst_oid))
-	  {
-	    repl_obj->set_key_value (key);
+	repl_obj->set_key_value(key);        
 
-	    append_repl_object (*repl_obj);
-	    er_log_repl_obj (repl_obj, "log_generator::set_key_to_repl_object");
+        /* Set the current transaction lsa. It may be rewritten later. */
+        repl_obj->set_lsa_stamp (*p_lsa);
 
-	    // remove
-	    (void) m_pending_to_be_added.erase (repl_obj_it);
+	append_repl_object(*repl_obj);
+	er_log_repl_obj(repl_obj, "log_generator::set_key_to_repl_object");
 
-	    found = true;
+	// remove
+	(void)m_pending_to_be_added.erase(repl_obj_it);
 
-	    break;
-	  }
+	found = true;
+
+	break;
       }
+    }
 
     if (!found)
-      {
-	assert (optional_recdes != NULL);
+    {
+      assert(optional_recdes != NULL);
+            
+      cubreplication::rec_des_row_repl_entry *entry =
+	new cubreplication::rec_des_row_repl_entry(cubreplication::repl_entry_type::REPL_UPDATE, class_name,
+	  *optional_recdes, *p_lsa);
 
-	cubreplication::rec_des_row_repl_entry *entry =
-		new cubreplication::rec_des_row_repl_entry (cubreplication::repl_entry_type::REPL_UPDATE, class_name,
-		    *optional_recdes);
+      entry->set_key_value(key);
 
-	append_repl_object (*entry);
+      append_repl_object(*entry);
 
-	er_log_repl_obj (entry, "log_generator::set_key_to_repl_object");
-      }
+      er_log_repl_obj(entry, "log_generator::set_key_to_repl_object");
+    }
 
-    free (class_name);
+    free(class_name);
   }
 
-  char *
-  log_generator::get_classname (const OID &class_oid)
+  void
+    log_generator::update_lsastamp_for_changed_repl_object (const OID &inst_oid)
   {
-    char *classname = NULL;
-    cubthread::entry *thread_p = &cubthread::get_entry ();
-    bool save = logtb_set_check_interrupt (thread_p, false);
-    if (heap_get_class_name (thread_p, &class_oid, &classname) != NO_ERROR || classname == NULL)
+    if (is_row_replication_disabled())
+    {
+      return;
+    }
+    
+    replication_object *repl_obj;
+    changed_attrs_row_repl_entry *changed_repl_entry;
+    cubthread::entry *thread_p = &cubthread::get_entry();
+    bool found = false;
+    int count_entries = (int) m_stream_entry.count_entries ();            
+    LOG_LSA *p_lsa = logtb_find_current_tran_lsa(thread_p);
+    assert (p_lsa != NULL);
+
+    for (int i = count_entries - 1; i >= 0; i--)
+    {
+      repl_obj = m_stream_entry.get_object_at(i);
+      changed_repl_entry = dynamic_cast<changed_attrs_row_repl_entry *> (repl_obj);
```

Comment:
I propose to use a virtual method in replication_object (with different behavior for class changed_attrs_row_repl_entry ) instead of dynamic_cast solution.